File: ios/TaskWraithKit/Sources/TaskWraithKit/RevealParams.swift
import Foundation

/// Swift twin of `src/shared/revealParams.ts` + `src/renderer/src/lib/advanceReveal.ts`.
///
/// The type-out / reveal cadence is shared across Electron and iOS so a user
/// switching Mac <-> phone sees the same feel. The TypeScript module is the
/// source of truth; THESE VALUES MUST MATCH IT. A drift-guard test
/// (`src/shared/revealParams.test.ts`) reads this file and fails CI if any
/// value diverges (same pattern as AppIconAvailability.swift).
///
/// The reveal is DISPLAY-ONLY: a length cursor over the byte-exact accumulated
/// string. It never writes back into the model's streaming text, so it cannot
/// drop or reorder content — the final rendered text always equals the full
/// content. Rates are TIME-BASED (chars/sec) so a 60Hz and a 120Hz ProMotion
/// device match. The cursor unit is the grapheme cluster — Swift `Character`
/// is grapheme-aware, so `Array(string)` already gives the right units and a
/// ZWJ emoji is never split.
public struct RevealParams: Sendable, Equatable {
    public var baseCharsPerSec: Double
    public var catchUpPerSec: Double
    public var maxCharsPerSec: Double
    public var maxCharsPerFrame: Int
    public var settleCharsPerSec: Double
    public var fadeTailChars: Int
    public var completeDrainMs: Double
    public var clearCeilingMs: Double
    public var coldSnapChars: Int

    public init(
        baseCharsPerSec: Double, catchUpPerSec: Double, maxCharsPerSec: Double,
        maxCharsPerFrame: Int, settleCharsPerSec: Double, fadeTailChars: Int,
        completeDrainMs: Double, clearCeilingMs: Double, coldSnapChars: Int
    ) {
        self.baseCharsPerSec = baseCharsPerSec
        self.catchUpPerSec = catchUpPerSec
        self.maxCharsPerSec = maxCharsPerSec
        self.maxCharsPerFrame = maxCharsPerFrame
        self.settleCharsPerSec = settleCharsPerSec
        self.fadeTailChars = fadeTailChars
        self.completeDrainMs = completeDrainMs
        self.clearCeilingMs = clearCeilingMs
        self.coldSnapChars = coldSnapChars
    }

    /// The cross-platform contract. Mirrors REVEAL_PARAMS in revealParams.ts.
    public static let shared = RevealParams(
        baseCharsPerSec: 55,
        catchUpPerSec: 3.0,
        maxCharsPerSec: 900,
        maxCharsPerFrame: 96,
        settleCharsPerSec: 170,
        fadeTailChars: 26,
        completeDrainMs: 250,
        clearCeilingMs: 300,
        coldSnapChars: 220
    )
}

/// smoothstep(t) = 3t^2 - 2t^3, clamped to [0,1].
public func revealSmoothstep(_ t: Double) -> Double {
    let x = min(1, max(0, t))
    return x * x * (3 - 2 * x)
}

/// Shared opacity-ramp curve. `distanceFromFrontier` is in graphemes back from
/// the reveal frontier: 0 = newest (≈0 opacity, fading in), `fadeTailChars` =
/// fully solid. Both platforms evaluate this so the fade easing matches.
public func revealFadeOpacity(
    _ distanceFromFrontier: Double, fadeTailChars: Int = RevealParams.shared.fadeTailChars
) -> Double {
    if fadeTailChars <= 0 { return 1 }
    return revealSmoothstep(distanceFromFrontier / Double(fadeTailChars))
}

/// Snap a candidate grapheme cursor FORWARD to the next word boundary so the
/// rendered slice ends on a whole word (a single-token span like `**bold**`
/// has no inner whitespace and so is revealed whole). Look-ahead capped at
/// `maxLookahead`; falls back to the raw index for a long unbroken run.
public func snapForwardToWordBoundary(
    _ graphemes: [Character], _ index: Int, _ maxLookahead: Int
) -> Int {
    if index >= graphemes.count { return graphemes.count }
    if index <= 0 { return max(0, index) }
    let limit = min(graphemes.count, index + max(0, maxLookahead))
    var j = index
    while j < limit {
        if graphemes[j].isWhitespace { return min(graphemes.count, j + 1) }
        j += 1
    }
    return index
}

/// The pure reveal-cursor step. Twin of `advanceReveal` in advanceReveal.ts.
/// Returns the next cursor (grapheme count into `next`), in [0, target].
/// Always converges; on a terminal reason drains within `completeDrainMs`.
public func advanceReveal(
    prev: String, next: String, revealed: Int, isComplete: Bool, dt: Double,
    params: RevealParams = .shared
) -> Int {
    let P = params
    let dtc = max(0, min(dt, 0.1))

    let nextChars = Array(next)
    let target = nextChars.count

    var rev = revealed
    if prev != next {
        let prevChars = Array(prev)
        let maxN = min(prevChars.count, nextChars.count)
        var common = 0
        while common < maxN && prevChars[common] == nextChars[common] { common += 1 }
        if rev > common { rev = common }
    }
    if rev < 0 { rev = 0 }
    if rev > target { rev = target }

    let backlog = target - rev
    if backlog <= 0 { return target }

    var rate = min(P.baseCharsPerSec + P.catchUpPerSec * Double(backlog), P.maxCharsPerSec)
    var stepCap = Double(P.maxCharsPerFrame)
    if isComplete {
        // target-based linear drain (not backlog-proportional, which never
        // reaches the end within the window), anti-slam bypassed.
        rate = max(rate, Double(target) / (P.completeDrainMs / 1000))
        stepCap = .infinity
    }

    let step = min(rate * dtc, stepCap)
    var nextRevealed = Int((Double(rev) + step).rounded(.down))
    if nextRevealed <= rev { nextRevealed = rev + 1 }
    if nextRevealed >= target { return target }

    return min(target, snapForwardToWordBoundary(nextChars, nextRevealed, P.fadeTailChars))
}

File: src/shared/revealParams.ts
/*
 * revealParams — the single source of truth for the type-out / reveal cadence,
 * shared by BOTH platforms. The Swift twin (ios/.../RevealParams.swift) mirrors
 * these EXACT values and is kept honest by a drift-guard test (same pattern as
 * src/shared/iconVariants.ts + its Swift twin).
 *
 * Design invariants (from the 13-agent design pass):
 *   - The reveal is DISPLAY-ONLY: a length cursor over the byte-exact
 *     accumulated string. It is strictly downstream of accumulation and NEVER
 *     writes back into chat records / chatByIdRef, so it cannot affect the
 *     token-drop gate (see lib/reconcileChatRefMap + streamingTokenDrop.test).
 *     The final rendered text always equals the full content by construction.
 *   - All rates are TIME-BASED (chars per SECOND), never per-frame, so a 60Hz
 *     Mac and a 120Hz ProMotion phone produce the same wall-clock cadence.
 *   - The cursor unit is the GRAPHEME cluster on both platforms
 *     (Intl.Segmenter('grapheme') on web / Swift `Character` on iOS), so the
 *     same `revealedLen` means the same text and a surrogate pair / ZWJ emoji
 *     is never split.
 *
 * This module is intentionally node-free and dependency-free.
 */

export interface RevealParams {
  /** Steady reveal floor (chars/sec) when there is little backlog. */
  baseCharsPerSec: number
  /** Proportional catch-up: rate grows by this per backlog-char/sec. */
  catchUpPerSec: number
  /** Hard ceiling (chars/sec) so a big burst never strobes. */
  maxCharsPerSec: number
  /**
   * Anti-slam clamp (max graphemes advanced in a single frame) for a resumed
   * backgrounded tab / suspended app that reports a huge dt. BYPASSED on the
   * terminal drain so a large final burst still finishes within completeDrainMs.
   */
  maxCharsPerFrame: number
  /** Rate (chars/sec) at which the fade tail solidifies while idle. */
  settleCharsPerSec: number
  /** Soft fade-band length, in graphemes, AND the word-boundary look-ahead cap. */
  fadeTailChars: number
  /** On any terminal reason, drain the remaining backlog within this window. */
  completeDrainMs: number
  /** Belt-and-suspenders: a stuck reveal can never pin a bubble open past this. */
  clearCeilingMs: number
  /**
   * Mid-run reattach / history-load snap threshold (graphemes). When content
   * is already present because the run predates this view, snap instant past
   * this size instead of animating. Scoped to reattach ONLY — a fresh run
   * always animates via the catch-up clock.
   */
  coldSnapChars: number
}

/** The cross-platform contract. The Swift twin mirrors these exact values. */
export const REVEAL_PARAMS: RevealParams = {
  baseCharsPerSec: 55,
  catchUpPerSec: 3.0,
  maxCharsPerSec: 900,
  maxCharsPerFrame: 96,
  settleCharsPerSec: 170,
  fadeTailChars: 26,
  completeDrainMs: 250,
  clearCeilingMs: 300,
  coldSnapChars: 220
}

/** smoothstep(t) = 3t^2 - 2t^3, clamped to [0,1]. */
export function smoothstep(t: number): number {
  const x = t < 0 ? 0 : t > 1 ? 1 : t
  return x * x * (3 - 2 * x)
}

/**
 * The SHARED opacity-ramp curve. Given a revealed grapheme's distance (in
 * graphemes) BACK from the reveal frontier, return its opacity in [0,1]:
 *   distance 0            → newest grapheme, fully faded-in-progress (~0)
 *   distance fadeTailChars → fully solid (1)
 * Both platforms evaluate THIS function so the fade easing matches — identical
 * cursor params alone are not enough (the feel is dominated by this curve).
 */
export function fadeOpacity(
  distanceFromFrontier: number,
  fadeTailChars: number = REVEAL_PARAMS.fadeTailChars
): number {
  if (fadeTailChars <= 0) return 1
  return smoothstep(distanceFromFrontier / fadeTailChars)
}
